components/script/dom/navigator.rs

```diff
@@ -81,9 +82,55 @@ impl Navigator {
         self.xr.get()
     }
 
-    pub fn gamepads(&self) -> DomRoot<GamepadList> {
-        self.gamepads
-            .or_init(|| GamepadList::new(&self.global(), &[]))
+    pub fn get_gamepad(&self, index: usize) -> Option<DomRoot<Gamepad>> {
+        self.gamepads.borrow().get(index).and_then(|g| g.get())
+    }
+
+    pub fn set_gamepad(&self, index: usize, gamepad: &Gamepad) {
+        if index >= self.gamepads.borrow().len() {
+            self.gamepads
+                .borrow_mut()
+                .resize_with(index + 1, Default::default);
+        }
+        if let Some(gamepad_to_set) = self.gamepads.borrow().get(index) {
+            gamepad_to_set.set(Some(gamepad));
+            if self.has_gamepad_gesture.get() {
+                gamepad.set_exposed(true);
+                if self.global().as_window().Document().is_fully_active() {
+                    gamepad.notify_event(GamepadEventType::Connected);
```

Comment:
Let's move firing the event to outside the borrow, because the JS code in the event handler could call `navigator.getGamePads`, and we'd get a double-borrow panic.